File: frontend/lib/services/health_service.dart
import 'package:omniconnect_fitness/services/api_client.dart';
import 'package:omniconnect_fitness/services/health_connect_service.dart';

/// DTO de resposta do endpoint de sync de saúde.
class HealthSyncResponse {
  final bool success;
  final String message;

  HealthSyncResponse({required this.success, required this.message});

  factory HealthSyncResponse.fromJson(Map<String, dynamic> json) =>
      HealthSyncResponse(
        success: json['success'] as bool? ?? true,
        message: json['message'] as String? ?? 'ok',
      );
}

/// Resumo de saúde retornado pelo GET /health/summary.
class HealthSummary {
  final double averageHeartRateBpm;
  final double activeCalories;
  final double totalCalories;
  final DateTime date;
  final bool isFromSmartwatch;
  final String smartwatchSourceName;

  HealthSummary({
    required this.averageHeartRateBpm,
    required this.activeCalories,
    required this.totalCalories,
    required this.date,
    this.isFromSmartwatch = false,
    this.smartwatchSourceName = '',
  });

  factory HealthSummary.fromJson(Map<String, dynamic> json) => HealthSummary(
        averageHeartRateBpm:
            (json['average_heart_rate_bpm'] as num?)?.toDouble() ?? 0,
        activeCalories: (json['active_calories'] as num?)?.toDouble() ?? 0,
        totalCalories: (json['total_calories'] as num?)?.toDouble() ?? 0,
        date: DateTime.parse(json['date'] as String),
        isFromSmartwatch: json['is_from_smartwatch'] as bool? ?? false,
        smartwatchSourceName: json['smartwatch_source_name'] as String? ?? '',
      );
}

/// Cliente HTTP para o módulo de saúde (FC + calorias).
class HealthService {
  final ApiClient _apiClient;

  HealthService({required ApiClient apiClient}) : _apiClient = apiClient;

  /// Sincroniza frequência cardíaca e calorias do dia com o backend.
  Future<HealthSyncResponse> syncHealthData({
    required List<HeartRateSample> heartRateSamples,
    required DailyCalories calories,
    required DateTime date,
  }) async {
    final body = <String, dynamic>{
      'date': _formatDate(date),
      'active_calories': calories.active,
      'total_calories': calories.total,
      'heart_rate_readings': heartRateSamples
          .map((s) => {
                'measured_at': s.measuredAt.toUtc().toIso8601String(),
                'bpm': s.bpm,
                'is_from_smartwatch': s.isFromSmartwatch,
                'source_name': s.sourceName,
              })
          .toList(),
    };

    return _apiClient.post<HealthSyncResponse>(
      '/health/sync',
      body: body,
      fromJson: (data) =>
          HealthSyncResponse.fromJson(data as Map<String, dynamic>),
    );
  }

  /// Busca o resumo de saúde do dia atual.
  Future<HealthSummary> getSummary({DateTime? date}) async {
    final query = <String, dynamic>{};
    if (date != null) query['date'] = _formatDate(date);

    return _apiClient.get<HealthSummary>(
      '/health/summary',
      queryParameters: query.isEmpty ? null : query,
      fromJson: (data) =>
          HealthSummary.fromJson(data as Map<String, dynamic>),
    );
  }

  String _formatDate(DateTime d) {
    final y = d.year.toString().padLeft(4, '0');
    final m = d.month.toString().padLeft(2, '0');
    final day = d.day.toString().padLeft(2, '0');
    return '$y-$m-$day';
  }
}

File: backend/app/services/health_service.py
"""Service de dados de saúde: frequência cardíaca e calorias."""

import logging
from datetime import date as date_type
from uuid import UUID

from sqlalchemy.ext.asyncio import AsyncSession

from app.dtos.health_dto import (
    HealthSyncRequestDTO,
    HealthSyncResponseDTO,
    HealthSummaryResponseDTO,
)
from app.repositories.health_repository import HealthRepository

logger = logging.getLogger(__name__)


class HealthService:
    """Serviço de sincronização e consulta de dados de saúde."""

    def __init__(self, session: AsyncSession):
        self.session = session
        self.repository = HealthRepository(session)

    async def sync_health_data(
        self, user_id: UUID, dto: HealthSyncRequestDTO
    ) -> HealthSyncResponseDTO:
        """Sincroniza amostras de FC e calorias do dia. Idempotente."""
        samples = [
            (reading.measured_at, reading.bpm, reading.is_from_smartwatch, reading.source_name)
            for reading in dto.heart_rate_readings
        ]
        saved_count = await self.repository.upsert_heart_rate_samples(user_id, samples)

        await self.repository.upsert_daily_calories(
            user_id=user_id,
            day=dto.date,
            active_calories=dto.active_calories,
            total_calories=dto.total_calories,
        )

        await self.repository.commit()

        logger.info(
            "Health sync: user=%s date=%s hr_samples=%d calories=%.1f",
            user_id,
            dto.date,
            saved_count,
            dto.active_calories,
        )

        return HealthSyncResponseDTO(
            success=True,
            message="Dados de saúde sincronizados com sucesso.",
            heart_rate_samples_saved=saved_count,
            date=dto.date,
        )

    async def get_summary(
        self, user_id: UUID, day: date_type
    ) -> HealthSummaryResponseDTO:
        """Retorna resumo de saúde do dia: FC média, min, max e calorias."""
        hr_stats = await self.repository.get_heart_rate_stats(user_id, day)
        calorie_log = await self.repository.get_daily_summary(user_id, day)

        return HealthSummaryResponseDTO(
            date=day,
            average_heart_rate_bpm=round(hr_stats["avg_bpm"], 1),
            min_heart_rate_bpm=hr_stats["min_bpm"],
            max_heart_rate_bpm=hr_stats["max_bpm"],
            heart_rate_samples=hr_stats["count"],
            active_calories=calorie_log.active_calories if calorie_log else 0.0,
            total_calories=calorie_log.total_calories if calorie_log else 0.0,
            is_from_smartwatch=hr_stats["is_from_smartwatch"],
            smartwatch_source_name=hr_stats["smartwatch_source_name"],
        )
